File: python-alert-service/messaging/consumer.py
"""RabbitMQ consumer for sensor update events using aio-pika (asyncio)."""

import asyncio
import json
import logging

import aio_pika

from metrics import server as metrics_server

logger = logging.getLogger("alert_service")


class AlertConsumer:
    """Consumes sensor.updated events from RabbitMQ via aio-pika.

    Blocking mode (worker_count=0): each message is evaluated inline in the
    consumer coroutine before the next message is fetched — sequential,
    strong at-least-once.

    Async mode (worker_count>0): messages are placed onto an asyncio.Queue;
    a pool of worker tasks drains it concurrently. The consumer acks
    immediately after enqueuing, so a crash between ack and evaluation
    drops the event — higher throughput at the cost of a weaker delivery
    guarantee. Backpressure: when the queue is full, put() awaits until
    a worker frees a slot.
    """

    def __init__(self, rabbitmq_url: str, callback, worker_count: int = 0):
        """Initialize the consumer.

        Args:
            rabbitmq_url: AMQP connection URL.
            callback: Async callable to invoke with each sensor event dict.
            worker_count: Number of worker tasks for async mode.
                          0 means blocking mode (no worker pool).
        """
        self.rabbitmq_url = rabbitmq_url
        self.callback = callback
        self.worker_count = worker_count

        if worker_count > 0:
            # Buffer = worker_count * 10 gives workers headroom without unbounded queuing.
            self._event_queue: asyncio.Queue | None = asyncio.Queue(
                maxsize=worker_count * 10
            )
            self._mode = "async"
        else:
            self._event_queue = None
            self._mode = "blocking"

    def start(self):
        """Schedule the consumer loop as an asyncio task."""
        loop = asyncio.get_event_loop()
        loop.create_task(self._run_workers_and_consumer())
        logger.info("Alert consumer started", extra={"mode": self._mode})

    async def _run_workers_and_consumer(self):
        """Start worker tasks (if async) then enter the consume loop."""
        if self._event_queue is not None:
            for _ in range(self.worker_count):
                asyncio.create_task(self._worker())

        while True:
            try:
                await self._consume()
            except Exception as e:
                logger.error(
                    "Consumer connection lost: %s — reconnecting in 5s", str(e)
                )
                await asyncio.sleep(5)

    async def _worker(self):
        """Pull events from the queue and evaluate them."""
        while True:
            event = await self._event_queue.get()
            try:
                await self.callback(event)
            except Exception:
                logger.exception("Worker failed to evaluate event: %s", event)
            finally:
                self._event_queue.task_done()

    async def _consume(self):
        """Connect to RabbitMQ and consume sensor.updated events."""
        connection = await aio_pika.connect_robust(self.rabbitmq_url)
        async with connection:
            channel = await connection.channel()

            # Prefetch limits how many unacked messages RabbitMQ sends at once.
            await channel.set_qos(prefetch_count=10)

            exchange = await channel.declare_exchange(
                "sensor_events", aio_pika.ExchangeType.FANOUT, durable=True
            )
            queue = await channel.declare_queue(
                "alert_service_python", durable=True
            )
            await queue.bind(exchange)

            logger.info("Connected to RabbitMQ, waiting for sensor events")

            async with queue.iterator() as queue_iter:
                async for message in queue_iter:
                    await self._on_message(message)

    async def _on_message(self, message: aio_pika.IncomingMessage):
        """Handle incoming message."""
        try:
            event = json.loads(message.body)
        except json.JSONDecodeError:
            logger.warning("Received invalid JSON message")
            await message.nack(requeue=False)
            return

        if event.get("event") == "sensor.updated":
            if metrics_server.collector is not None:
                metrics_server.collector.inc_received()
            logger.info(
                "Received sensor.updated event",
                extra={
                    "sensor_id": event.get("sensor_id"),
                    "value": event.get("value"),
                    "trace_id": event.get("trace_id"),
                },
            )
            if self._event_queue is not None:
                # Async: enqueue for worker pool; awaits if queue is full (backpressure).
                await self._event_queue.put(event)
            else:
                # Blocking: evaluate inline before acking.
                await self.callback(event)

        await message.ack()

File: go-alert-service/messaging/consumer.go
package messaging

import (
	"encoding/json"
	"log/slog"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"

	"iot-alert-service/metrics"
)

// SensorEvent represents a sensor.updated event from RabbitMQ.
type SensorEvent struct {
	Event     string  `json:"event"`
	SensorID  string  `json:"sensor_id"`
	Value     float64 `json:"value"`
	Type      string  `json:"type"`
	Unit      string  `json:"unit"`
	Timestamp string  `json:"timestamp"`
	TraceID   string  `json:"trace_id"`
}

// AlertConsumer consumes sensor.updated events from RabbitMQ.
type AlertConsumer struct {
	url      string
	callback func(SensorEvent)
	// async pipeline fields; nil when running in blocking mode
	eventCh chan SensorEvent
}

// NewAlertConsumer creates a new alert consumer in blocking mode.
func NewAlertConsumer(url string, callback func(SensorEvent)) *AlertConsumer {
	return &AlertConsumer{url: url, callback: callback}
}

// NewAsyncAlertConsumer creates a consumer that dispatches events to a
// buffered worker pool. workerCount goroutines drain the channel; when the
// channel is full the consumer blocks (backpressure).
func NewAsyncAlertConsumer(url string, callback func(SensorEvent), workerCount int) *AlertConsumer {
	// Buffer = workerCount * 10 gives workers headroom without unbounded queuing.
	ch := make(chan SensorEvent, workerCount*10)
	c := &AlertConsumer{url: url, callback: callback, eventCh: ch}
	for i := 0; i < workerCount; i++ {
		go c.worker()
	}
	return c
}

func (c *AlertConsumer) worker() {
	for event := range c.eventCh {
		func() {
			defer func() {
				if r := recover(); r != nil {
					slog.Error("Worker panic recovered", "error", r, "sensor_id", event.SensorID)
				}
			}()
			c.callback(event)
		}()
	}
}

// Start begins consuming in a background goroutine with reconnect logic.
func (c *AlertConsumer) Start() {
	mode := "blocking"
	if c.eventCh != nil {
		mode = "async"
	}
	go c.consumeLoop()
	slog.Info("Alert consumer started", "mode", mode)
}

func (c *AlertConsumer) consumeLoop() {
	for {
		err := c.consume()
		if err != nil {
			slog.Error("Consumer connection lost — reconnecting in 5s", "error", err.Error())
			time.Sleep(5 * time.Second)
		}
	}
}

func (c *AlertConsumer) consume() error {
	conn, err := amqp.Dial(c.url)
	if err != nil {
		return err
	}
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil {
		return err
	}
	defer ch.Close()

	err = ch.ExchangeDeclare("sensor_events", "fanout", true, false, false, false, nil)
	if err != nil {
		return err
	}

	q, err := ch.QueueDeclare("alert_service_go", true, false, false, false, nil)
	if err != nil {
		return err
	}

	err = ch.QueueBind(q.Name, "", "sensor_events", false, nil)
	if err != nil {
		return err
	}

	// Prefetch limits how many unacked messages RabbitMQ sends at once.
	err = ch.Qos(10, 0, false)
	if err != nil {
		return err
	}

	// autoAck=false: ack only after successful processing so messages are
	// not lost if the evaluator crashes mid-flight.
	msgs, err := ch.Consume(q.Name, "", false, false, false, false, nil)
	if err != nil {
		return err
	}

	slog.Info("Connected to RabbitMQ, waiting for sensor events")

	for msg := range msgs {
		var event SensorEvent
		if err := json.Unmarshal(msg.Body, &event); err != nil {
			slog.Warn("Received invalid JSON message", "error", err.Error())
			// Nack without requeue — malformed messages cannot be retried usefully.
			msg.Nack(false, false)
			continue
		}
		if event.Event == "sensor.updated" {
			metrics.EventsReceived.Add(1)
			slog.Info("Received sensor.updated event",
				"sensor_id", event.SensorID, "value", event.Value, "trace_id", event.TraceID)
			if c.eventCh != nil {
				// Async: send to worker pool; blocks if channel is full (backpressure).
				c.eventCh <- event
			} else {
				// Blocking: evaluate inline before acking.
				c.callback(event)
			}
		}
		msg.Ack(false)
	}

	return nil
}
